src/FirmwarePlugin/APM/ArduCopterFirmwarePlugin.cc

```diff
@@ -136,3 +136,44 @@ bool ArduCopterFirmwarePlugin::vehicleYawsToNextWaypointInMission(const Vehicle*
     return true;
 }
 
+void ArduCopterFirmwarePlugin::guidedModeGotoLocation(Vehicle* vehicle, const QGeoCoordinate& gotoCoord)
+{
+    if (qIsNaN(vehicle->altitudeRelative()->rawValue().toDouble())) {
+        qgcApp()->showMessage(QStringLiteral("Unable to go to location, vehicle position not known."));
+        return;
+    }
+
+    setGuidedMode(vehicle, true);
+
+    if (vehicle->capabilityBits() & MAV_PROTOCOL_CAPABILITY_COMMAND_INT) {
+
+        mavlink_message_t msg;
+        mavlink_set_position_target_global_int_t cmd;
+
+        memset(&cmd, 0, sizeof(cmd));
+
+        cmd.target_system    = static_cast<uint8_t>(vehicle->id());
+        cmd.target_component = static_cast<uint8_t>(vehicle->defaultComponentId());
+        cmd.coordinate_frame = MAV_FRAME_GLOBAL_RELATIVE_ALT;
+        cmd.type_mask = 0x0DF8; // Position Only
+        cmd.lat_int = static_cast<int>(gotoCoord.latitude()*1e7);
+        cmd.lon_int = static_cast<int>(gotoCoord.longitude()*1e7);
+        cmd.alt = vehicle->altitudeRelative()->rawValue().toFloat();
```

Comment:
This isn't really related to this implementation in particular: can't QGeoCoordinate store either a 2D or 3D position?  If this method is called with a 3D position, should this really be the behaviour?  It seems pretty non-intuitive.

Reply:
The coord passed in does not have an altitude in it. Just lat/lon is provided.

Reply:
Is that just a function of the fact that right now, the only way of setting a guided mode target is by clicking on the map though?  What if I wanted to implement some other UI element which provided a 3D coordinate?

Reply:
> What if I wanted to implement some other UI element which provided a 3D coordinate?

It's a function of the current ui interface (which doesn't allow altitude change on goto) as well as what is expected from the FirmwarePlugin method. They were implemented at the same time. Although totally possible to add from ui and interface side and folks have been asking for altitude change in combination with goto location anyway.

Reply:
So, it might be ok then to change the implementations such that if the coordinate supplied is 3D, it sets the altitude as requested, but if the coordinate is just 2D, it uses the current altitude (as is the present behaviour)?

Reply:
Sure but that will require the corresponding ui for it as well. I would put that in a separate pull.

Reply:
@tilaktilak To implement that you can look at how the Guided Pause command always shows the altitude slider and pulls values from it.